File: sdk/luma/hub.py
from typing import Dict, Optional

from ._http import Http


class HubClient:
    def __init__(self, http: Http, namespace: str):
        self._http = http
        self._base = f"/v1/db/{namespace}"

    def ingest(self, text: str, id: Optional[str] = None,
               metadata: Optional[Dict] = None) -> dict:
        body: Dict = {"text": text}
        if id:
            body["id"] = id
        if metadata:
            body["metadata"] = metadata
        return self._http.post(f"{self._base}/ingest", body)

    async def aingest(self, text: str, id: Optional[str] = None,
                      metadata: Optional[Dict] = None) -> dict:
        body: Dict = {"text": text}
        if id:
            body["id"] = id
        if metadata:
            body["metadata"] = metadata
        return await self._http.apost(f"{self._base}/ingest", body)

    def search(self, query: str, sql_filter: Optional[str] = None, limit: int = 10) -> dict:
        body: Dict = {"query": query, "limit": limit}
        if sql_filter:
            body["sql_filter"] = sql_filter
        return self._http.post(f"{self._base}/search", body)

    async def asearch(self, query: str, sql_filter: Optional[str] = None, limit: int = 10) -> dict:
        body: Dict = {"query": query, "limit": limit}
        if sql_filter:
            body["sql_filter"] = sql_filter
        return await self._http.apost(f"{self._base}/search", body)

File: sdk/typescript/src/hub.ts
import { HttpClient } from "./http.js";
import type {
  HubIngestOptions,
  HubSearchOptions,
  HubSearchResult,
  Metadata,
} from "./types.js";

/**
 * HubClient — LumaDatabase hybrid layer for a single namespace.
 * Handles auto-chunking, embedding generation, and hybrid SQL+vector search.
 *
 * Obtain via `client.db(namespace)`.
 */
export class HubClient {
  private readonly base: string;

  constructor(
    private readonly http: HttpClient,
    namespace: string,
  ) {
    this.base = `/v1/db/${encodeURIComponent(namespace)}`;
  }

  /**
   * Ingest a text document. The server chunks it, generates embeddings, and
   * stores both the vector and the raw text.
   */
  ingest(
    text: string,
    opts: HubIngestOptions = {},
  ): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { text };
    if (opts.id !== undefined) body["id"] = opts.id;
    if (opts.metadata !== undefined) body["metadata"] = opts.metadata;
    return this.http.post(`${this.base}/ingest`, body);
  }

  /**
   * Hybrid search: SQL pre-filter followed by vector similarity ranking.
   */
  search(
    query: string,
    opts: HubSearchOptions = {},
  ): Promise<HubSearchResult> {
    const body: Record<string, unknown> = { query, limit: opts.limit ?? 10 };
    if (opts.sql_filter !== undefined) body["sql_filter"] = opts.sql_filter;
    return this.http.post(`${this.base}/search`, body);
  }
}
